src/common/buffer.cc

```diff
@@ -2268,47 +2268,46 @@ int buffer::list::write_fd(int fd) const
   ssize_t bytes = 0;
 
   std::list<ptr>::const_iterator p = _buffers.begin();
-  while (p != _buffers.end() && iovlen != IOV_MAX) {
+  uint64_t size = MIN( _buffers.size(), IOV_MAX);
+  while (iovlen < size) {
```

Comment:
This version silently truncates written bufferlists to IOV_MAX - (bufferptrs of size = 0), because once the IOV_MAX limit is hit, the code doesn't reiterate over the rest of bufferlist.